server/drivers/futaba.c

```diff
@@ -118,6 +126,65 @@ MODULE_EXPORT int stay_in_foreground = 0;
 MODULE_EXPORT int supports_multiple = 0;
 MODULE_EXPORT char *symbol_prefix = "futaba_";
 
+static void thread_timer(void *params)
+
+    /* This thread loops continually. If p->flash is set then the colon icon(s)
+     * on the display will flash once per second approximately. Trying to send
+     * the report BEFORE the driver is fully initialised causes a crash in
+     * libusb, hence the need for the driver_init_complete flag */
+{
+    futabaReport_t my_report;
+    Driver *drvthis;
+    PrivateData *p;
+
+    drvthis = (Driver *) params;
+    p = (PrivateData *) drvthis->private_data;
+
+    while (1) {
+        if (p->flash) {
+            my_report.opcode = FUTABA_OPCODE_SYMBOL;
+            my_report.param1 = 0x02;
+            my_report.type.sym.count = 1;
+            my_report.type.sym.symbol[0].symName = 0x30;
+            my_report.type.sym.symbol[0].state = FUTABA_SYM_ON;
+            futaba_send_report(p->my_handle, &my_report);
+            if (p->long_time) { // turn on second colon if
+                // time is HH:MM:SS
+                my_report.opcode = FUTABA_OPCODE_SYMBOL;
+                my_report.param1 = 0x02;
+                my_report.type.sym.count = 1;
+                my_report.type.sym.symbol[0].symName = 0x34;
+                my_report.type.sym.symbol[0].state = FUTABA_SYM_ON;
+                futaba_send_report(p->my_handle, &my_report);
+            }
+            p->flash_state = 1;
+            sleep(1);
+        }
+        if (p->driver_init_complete && p->flash_state) {
+            // only turn the colon off if driver is
+            // initialised AND colon is lit
+            // otherwise the display will flash intermittently
+            // when scrolling text
+            my_report.opcode = FUTABA_OPCODE_SYMBOL;
+            my_report.param1 = 0x02;
+            my_report.type.sym.count = 1;
+            my_report.type.sym.symbol[0].symName = 0x30;
+            my_report.type.sym.symbol[0].state = FUTABA_SYM_OFF;
+            futaba_send_report(p->my_handle, &my_report);
+            my_report.opcode = FUTABA_OPCODE_SYMBOL;
+            my_report.param1 = 0x02;
+            my_report.type.sym.count = 1;
+            my_report.type.sym.symbol[0].symName = 0x34;
+            my_report.type.sym.symbol[0].state = FUTABA_SYM_OFF;
+            futaba_send_report(p->my_handle, &my_report);
+            p->flash_state = 0;
+            sleep(1);
+        }
+    }
+
```

Comment:
When none of options `p->flash` nor `p->flash_state` will be set, thread will consume 100% of processor, as there is no sleep at all.

Reply:
Indeed.  However, as it effectively does nothing in the loop it doesn't appear to affect the machine at all.  It has been running on my rig continuously for 5 days 24/7 and with a passively cooled pentium4, max temp has been 40C only (Max is 115C).  Note that Kodi is also running continuously.

I can't see a different way of blinking those icons within the driver without using a separate thread to do it.  If there is a better way of doing it, I'd be pleased to implement it.  BTW, lis driver also implements a separate thread (which reads not writes) but has no sleep() in it at all.

Reply:
The reason why this works in lis is because it uses a blocking call IIRC. Here I don't see a blocking call but I guess on your setup just is so that a flag is set and you have a sleep() in your tests most of the time.

That lis pulls in libpthread is a bit annoing but it is already there, so little point in trying to avoid it here. However it might be possible to implement blinking without a extra thread and at the same time fix some other issues with the code. I'm writing a proper review about that ...